stablehlo/reference/InterpreterApi.cpp

```diff
@@ -0,0 +1,215 @@
+/* Copyright 2023 The StableHLO Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "stablehlo/reference/InterpreterApi.h"
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Errc.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/SourceMgr.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Support/DebugStringHelper.h"
+#include "stablehlo/reference/Errors.h"
+#include "stablehlo/reference/InterpreterOps.h"
+#include "stablehlo/reference/NumPy.h"
+#include "stablehlo/reference/Ops.h"
+#include "stablehlo/tests/CheckOps.h"
```

Comment:
I don't think CheckOps should ship with the default implementation.

I believe the default fallback should just be "InterpreterOps", and should allow subclassing to support InterpreterOps + their own dialect ops / custom call ops.

Could have methods for:
- `handleInterpreterOp` (this should be marked final, not overridable)
- `handleUnregisteredOp` (check dialect could be added here)
- `handleCustomCallOp` (any additional handling)

Or could be a map of `op_name -> fallbackHandler`, then users could add their own `addCheckDialectFallbacks(handler)` methods to populate the fallback handler class.

But for now I think `CheckOps` should remain a feature of `StablehloTranslateMain`, as exposing them to the wild has broader maintenance implications.

Reply:
Hm, how about just implementing the `InterpreterOp` in the default fallback functor in lieu of 3 separate callbacks? And then leaving the `CheckOp` kernels within the impl of `StablehloTranslateMain`?